### PR TITLE
Add tlsVerify bool to SearchImage for varlink

### PR DESCRIPTION
**Pull request by rhatdan**
Cockpit wants to be able to search images on systems without
tlsverify turned on.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Fixes https://github.com/containers/libpod/issues/2293

**Comment by baude** [comment]
this needs a `make varlink_generate_api_generate` to update the api documentation; after that, LGTM

**Comment by KKoukiou** [comment]
When the registry container the image we are searching for is listed in insecure registries section in conf file and tlsVerify: true is passed to the API, there is not error but the image is returned.
Is this expected?

According to the cli docs for tlsVerify this should not happen, https://github.com/containers/libpod/blob/master/docs/podman-search.1.md

**Comment by rhatdan** [comment]
If you manually set the flsVerify flag to true and the registry is listed as insecure.  Then the containers/image should only use the registry if it can be verified.  IE it will connect with https only not fall back to http://.  At least that is my understanding.  @umohnani8 @vrothberg @mtrmac Please verify.

**Comment by KKoukiou** [comment]
> If you manually set the flsVerify flag to true and the registry is listed as insecure. Then the containers/image should only use the registry if it can be verified. IE it will connect with https only not fall back to http://. At least that is my understanding. @umohnani8 @vrothberg @mtrmac Please verify.

Correct. This behaviour is expected according to the docs and also the podman CLI follows this behaviour.
However, the varlink API does not respect the tlsVerify parameter if the registry is listed in insecure section, but instead it always returns the images from insecure registries.

**Comment by rhatdan** [comment]
@KKoukiou You are saying with this patch varlink is working differently then podman search --tlsverify=true?

**Comment by KKoukiou** [comment]
@rhatdan Correct. 
```
$ sudo varlink call unix:/run/podman/io.podman/io.podman.SearchImage '{"name": "my-ubuntu", "tlsVerify":true}'
{
  "images": [
    {
      "description": "",
      "is_automated": false,
      "is_official": false,
      "name": "my-ubuntu",
      "star_count": 0
    }
  ]
}
$ sudo podman search my-ubuntu --tls-verify=true
ERRO[0000] error searching registry "localhost:5000": couldn't search registry "localhost:5000": pinging docker registry returned: Get https://localhost:5000/v2/: http: server gave HTTP response to HTTPS client 
```
my ubuntu image exists in a local private registry, and the configuration looks like this.
```
$ sudo cat /etc/containers/registries.conf 
[registries.search]
registries = ['localhost:5000']

[registries.insecure]
registries = ['localhost:5000']
```

**Comment by rhatdan** [comment]
Ok then that is a bug.

**Comment by rhatdan** [comment]
I think we have an issue with the Varlink API, in that we have three settings.

One Use Default TLSVerify (Whatever is set in registries.conf)
Force on, force off.

In the CLI we can check if the user specified a tlsverify flag and then set it, otherwise we use default.

In the API, how should we indicate this. Perhaps have three contants.  Default, On, Off?

**Comment by mheon** [comment]
There was talk of an OptionalBool type (unset, true, false) elsewhere

**Comment by rhatdan** [comment]
Yes we are using that, but I don't see how we do that in Varlink.

**Comment by rhatdan** [comment]
@baude Does varlink have optional parameters?

**Comment by baude** [comment]
> @baude Does varlink have optional parameters?

not per'se ...

**Comment by jwhonce** [comment]
@baude A nullable bool aka ?bool should work, yes?  Use nil for Default and then true or false if set by user.

**Comment by baude** [comment]
you saw this in varlink docs as valid ?

**Comment by jwhonce** [comment]
@baude Correct. Search https://varlink.org/Interface-Definition for nullable.

**Comment by rhatdan** [comment]
@jwhonce @baude I have made tlsVerify optionalBool,  Not sure how you call this from Varlink, but I think this allows the passing of the optionalBool.

**Comment by rhatdan** [comment]
@jwhonce @baude PTAL

**Comment by baude** [comment]
LGTM, lets give it a try!

File: pkg/varlinkapi/images.go [hunk 1]
```diff
@@ -346,8 +346,8 @@ func (i *LibpodAPI) PushImage(call iopodman.VarlinkCall, name, tag string, tlsVe
 		DockerRegistryCreds: registryCreds,
 		DockerCertPath:      certDir,
 	}
-	if !tlsVerify {
-		dockerRegistryOptions.DockerInsecureSkipTLSVerify = types.OptionalBoolTrue
+	if tlsVerify != nil {
+		dockerRegistryOptions.DockerInsecureSkipTLSVerify = types.NewOptionalBool(*tlsVerify)
```
Comment (mtrmac) [inline pkg/varlinkapi/images.go]: !! This reverses the semantics!!

File: pkg/varlinkapi/images.go [hunk 2]
```diff
@@ -450,8 +450,11 @@ func (i *LibpodAPI) RemoveImage(call iopodman.VarlinkCall, name string, force bo
 
 // SearchImages searches all registries configured in /etc/containers/registries.conf for an image
 // Requires an image name and a search limit as int
-func (i *LibpodAPI) SearchImages(call iopodman.VarlinkCall, query string, limit *int64) error {
+func (i *LibpodAPI) SearchImages(call iopodman.VarlinkCall, query string, limit *int64, tlsVerify *bool) error {
 	sc := image.GetSystemContext("", "", false)
+	if tlsVerify != nil {
+		sc.DockerInsecureSkipTLSVerify = types.NewOptionalBool(*tlsVerify)
```
Comment (mtrmac) [inline pkg/varlinkapi/images.go]: !! This reverses the semantics!!

File: pkg/varlinkapi/images.go [hunk 3]
```diff
@@ -609,8 +612,8 @@ func (i *LibpodAPI) PullImage(call iopodman.VarlinkCall, name string, certDir, c
 		DockerRegistryCreds: registryCreds,
 		DockerCertPath:      certDir,
 	}
-	if tlsVerify {
-		dockerRegistryOptions.DockerInsecureSkipTLSVerify = types.NewOptionalBool(!tlsVerify)
+	if tlsVerify != nil {
+		dockerRegistryOptions.DockerInsecureSkipTLSVerify = types.NewOptionalBool(*tlsVerify)
```
Comment (mtrmac) [inline pkg/varlinkapi/images.go]: !! This reverses the semantics!!

File: pkg/varlinkapi/images.go [hunk 4]
```diff
@@ -653,8 +656,8 @@ func (i *LibpodAPI) ContainerRunlabel(call iopodman.VarlinkCall, input iopodman.
 	dockerRegistryOptions := image.DockerRegistryOptions{
 		DockerCertPath: input.CertDir,
 	}
-	if !input.TlsVerify {
-		dockerRegistryOptions.DockerInsecureSkipTLSVerify = types.OptionalBoolTrue
+	if input.TlsVerify != nil {
+		dockerRegistryOptions.DockerInsecureSkipTLSVerify = types.NewOptionalBool(*input.TlsVerify)
```
Comment (mtrmac) [inline pkg/varlinkapi/images.go]: !! This reverses the semantics!!

File: pkg/varlinkapi/images.go (hunk 1 above)
Comment (rhatdan) [reply pkg/varlinkapi/images.go]: Fixed

File: pkg/varlinkapi/images.go (hunk 2 above)
Comment (rhatdan) [reply pkg/varlinkapi/images.go]: Fixed

File: pkg/varlinkapi/images.go (hunk 3 above)
Comment (rhatdan) [reply pkg/varlinkapi/images.go]: Fixed

File: pkg/varlinkapi/images.go (hunk 4 above)
Comment (rhatdan) [reply pkg/varlinkapi/images.go]: Fixed

**Comment by rhatdan** [comment]
@mtrmac It is fine with me to change the semantics, @baude @jwhonce WDYT?

**Comment by jwhonce** [comment]
LGTM with respect to the API.  This will break the API tests in python-podman.

**Comment by rhatdan** [comment]
Yes @jwhonce, can you give me a /lgtm

**Pull request merged**